core/src/main/java/org/apache/iceberg/deletes/RoaringPositionBitmap.java

```diff
@@ -0,0 +1,309 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.deletes;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.List;
+import java.util.function.LongConsumer;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.roaringbitmap.RoaringBitmap;
+
+/**
+ * A bitmap that supports positive 64-bit positions, but is optimized for cases where most positions
+ * fit in 32 bits by using an array of 32-bit Roaring bitmaps. The bitmap array is grown as needed
+ * to accommodate the largest value.
+ *
+ * <p>Incoming 64-bit positions are divided into a 32-bit "key" using the most significant 4 bytes
+ * and a 32-bit position using the least significant 4 bytes. For each key in the set of positions,
+ * a 32-bit Roaring bitmap is maintained to store a set of 32-bit positions for that key.
+ *
+ * <p>To test whether a certain position is set, its most significant 4 bytes (the key) are used to
+ * find a 32-bit bitmap and the least significant 4 bytes are tested for inclusion in the bitmap. If
+ * a bitmap is not found for the key, then the position is not set.
+ *
+ * <p>Note that positions must be greater than or equal to 0 add less than {@link #MAX_POSITION}.
+ * This bitmap does not support positions larger than {@link #MAX_POSITION} as the bitmap array
+ * length is a signed 32-bit integer that must be greater than or equal to 0.
+ */
+class RoaringPositionBitmap {
+
+  static final long MAX_POSITION = toPosition(Integer.MAX_VALUE - 1, Integer.MIN_VALUE);
+  private static final RoaringBitmap[] EMPTY_BITMAP_ARRAY = new RoaringBitmap[] {};
+  private static final long BITMAP_COUNT_SIZE_BYTES = 8L;
+  private static final long BITMAP_KEY_SIZE_BYTES = 4L;
+
+  private RoaringBitmap[] bitmaps;
+
+  RoaringPositionBitmap() {
+    this.bitmaps = EMPTY_BITMAP_ARRAY;
+  }
+
+  private RoaringPositionBitmap(RoaringBitmap[] bitmaps) {
+    this.bitmaps = bitmaps;
+  }
+
+  /**
+   * Adds a position to the bitmap.
+   *
+   * @param pos the row position
+   */
+  public void add(long pos) {
+    validatePosition(pos);
+    int high = highBytes(pos);
+    int low = lowBytes(pos);
+    allocateBitmapsIfNeeded(high + 1 /* required number of bitmaps */);
+    bitmaps[high].add(low);
+  }
+
+  /**
+   * Adds a range of positions to the bitmap.
+   *
+   * @param posStartInclusive the start position of the range (inclusive)
+   * @param posEndExclusive the end position of the range (exclusive)
+   */
+  public void addRange(long posStartInclusive, long posEndExclusive) {
```

Comment:
This can be optimized in the future but we have no use case in the repo beyond tests.

Reply:
Should we make this package-private then? This has implications for run encoding in the bitmaps.

Reply:
Our public `PositionDeleteIndex` where we will eventually consume this class actually supports marking ranges as removed. Therefore, it is needed, but we shouldn't be concerned optimizing this right away.